# <a target="_blank" href="https://miptstats.github.io/courses/ad_mipt.html">Phystech@DataScience</a>


In [1]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import seaborn as sns
sns.set(font_scale=1.3)

### Линейная регрессия: практика
Перед началом работы догрузим необходимые нам функции



In [4]:
from sklearn.datasets import load_diabetes
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

Рассмотрим данные исследования 2004 года о зависимости риска развития диабета от пола, возраста, индекса массы тела, среднего кровяного давления и других показателей. 



#### Загрузка данных

In [6]:
data = load_diabetes()

Функция `sklearn.datasets.load_diabetes()` возвращает словарь. В поле `data` записана матрица регрессоров, в которой данные предварительно центрированы и нормированы. В поле `target` записана мера прогрессирования заболевания в течении года. В поле `DESCR` можно прочитать подробнее о данных.

Посмотрим на описание датасета.

In [7]:
print(data['DESCR'])

.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

:Number of Instances: 442

:Number of Attributes: First 10 columns are numeric predictive values

:Target: Column 11 is a quantitative measure of disease progression one year after baseline

:Attribute Information:
    - age     age in years
    - sex
    - bmi     body mass index
    - bp      average blood pressure
    - s1      tc, total serum cholesterol
    - s2      ldl, low-density lipoproteins
    - s3      hdl, high-density lipoproteins
    - s4      tch, total cholesterol / HDL
    - s5      ltg, possibly log of serum triglycerides level
    - s6      glu, blood sugar level

Note: Each of these 10 feature variables have bee

Поле `data` содержит матрицу размера 442 $\times$ 10, где 442 — количество пациентов, а 10 — количество признаков (возраст, пол, и т.д.). 
Строки матрицы соответствуют пациентам, столбцы — признакам.

In [8]:
data['data'].shape

(442, 10)

Целевая переменная $-$ мера прогрессирования заболевания в течении года.

In [9]:
data['target'].shape

(442,)

Создайте матрицу регрессоров $X$ (data) и столбец наблюдений $y$ (целевая переменная).

In [11]:
X, y = (data['data'], data['target'])

#### Обучение моделей

Разбейте данные случайно на две части — обучающую и тестовую в соотношении 80:20.

Если что-то забыли или что-то не понятно, можно почитать справку: https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html

При разбиении датасета стоит зафиксировать случайность для воспроизводимости результатов, поставив `random_state=42` в функцию разбиения (можете ставить любое число, необязательно именно 42)

In [19]:
X_train, X_test = train_test_split(X, test_size = 0.2, random_state=42)
y_train, y_test = train_test_split(y, test_size = 0.2, random_state=42)
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(353, 10) (89, 10) (353,) (89,)


Заведите модель линейной регрессии из `sklearn` и обучите ее по обучающей части данных.

In [22]:
model = LinearRegression(fit_intercept=True)
model.fit(X_train, y_train)

LinearRegression()

Посмотрите на результат обучения. Выведите коэффициенты перед признаками и свободный коэффициент.

In [23]:
model.coef_

array([  37.90402135, -241.96436231,  542.42875852,  347.70384391,
       -931.48884588,  518.06227698,  163.41998299,  275.31790158,
        736.1988589 ,   48.67065743])

Выполните предсказание модели на тестовой выборке.

In [25]:
y_pred = model.predict(X_test)
y_pred

array([139.5475584 , 179.51720835, 134.03875572, 291.41702925,
       123.78965872,  92.1723465 , 258.23238899, 181.33732057,
        90.22411311, 108.63375858,  94.13865744, 168.43486358,
        53.5047888 , 206.63081659, 100.12925869, 130.66657085,
       219.53071499, 250.7803234 , 196.3688346 , 218.57511815,
       207.35050182,  88.48340941,  70.43285917, 188.95914235,
       154.8868162 , 159.36170122, 188.31263363, 180.39094033,
        47.99046561, 108.97453871, 174.77897633,  86.36406656,
       132.95761215, 184.53819483, 173.83220911, 190.35858492,
       124.4156176 , 119.65110656, 147.95168682,  59.05405241,
        71.62331856, 107.68284704, 165.45365458, 155.00975931,
       171.04799096,  61.45761356,  71.66672581, 114.96732206,
        51.57975523, 167.57599528, 152.52291955,  62.95568515,
       103.49741722, 109.20751489, 175.64118426, 154.60296242,
        94.41704366, 210.74209145, 120.2566205 ,  77.61585399,
       187.93203995, 206.49337474, 140.63167076, 105.59

Реализуйте метрики MSE, MAE, MAPE без использования `sklearn` и других готовых реализаций.

Пусть $Y_1, ..., Y_n$ &mdash; истинные значения, а $\widehat{Y}_1, ..., \widehat{Y}_n$ &mdash; предсказания.


Метрика MSE (mean squared error) определяется как
$$MSE = \frac{1}{n}\sum_{i=1}^n \left(Y_i - \widehat{Y}_i\right)^2.$$

Метрика **MAE** (*mean absolute error*), определяемая как 
$$MAE = \frac{1}{n}\sum_{i=1}^n \left|Y_i - \widehat{Y}_i\right|.$$

Метрика **MAPE** (*mean absolute percentage error*), определяемая как 
$$MAPE = 100\% \cdot \frac{1}{n}\sum_{i=1}^n \frac{\left|Y_i - \widehat{Y}_i\right|}{Y_i}.$$


In [36]:
def mean_squared_error(y_true, y_pred):
    return np.mean((y_pred - y_true)**2)

def mean_absolute_error(y_true, y_pred):
    return np.mean(np.abs(y_pred-y_true))

def mean_absolute_percentage_error(y_true, y_pred):
    return 100*np.mean(1/y_true*np.abs(y_pred-y_true))

Посчитайте MSE, MAE, MAPE на тестовой выборке и выведите с точностью до трех знаков после запятой.

In [40]:
np.round(mean_squared_error(y_test, y_pred), 3)


2900.194

In [41]:
np.round(mean_absolute_error(y_test, y_pred), 3)

42.794

In [46]:
np.round(mean_absolute_percentage_error(y_test, y_pred), 3)


37.5

### Различные разбиения

Разбейте данные на тренировочную и тестовую выборки еще раз: измените пераметр `random_state`. Для каждого из разбиений обучите модель и получите метрики качества. Меняются ли эти метрики, если менять разбиение данных?

In [76]:
mse = []
mae = []
mape = []
for test_size in [0.2, 0.3, 0.4]:
    for state in [1, 495, 324]:
    
        # разбиение данных
        X_train, X_test = train_test_split(X, test_size = test_size, random_state = state)
        y_train, y_test = train_test_split(y, test_size = test_size, random_state = state)
        
        # обучение модели
        model = LinearRegression(fit_intercept=True)
        model.fit(X_train, y_train)

        # предсказание модели
        y_pred = model.predict(X_test)

        # подсчет метрики на тестовой выборке
        mse.append(np.round(mean_squared_error(y_test, y_pred), 3))
        mae.append(np.round(mean_absolute_error(y_test, y_pred), 3))
        mape.append(np.round(mean_absolute_percentage_error(y_test, y_pred), 3))

In [77]:
mse, mae, mape

([2992.581,
  2794.888,
  3010.615,
  2827.084,
  2815.377,
  2777.518,
  2962.953,
  3023.882,
  2976.654],
 [41.975, 41.751, 45.098, 41.649, 42.834, 43.511, 43.058, 44.757, 44.956],
 [32.494, 38.734, 41.314, 33.971, 40.285, 39.909, 36.591, 42.237, 38.161])

**Вывод:**

От random_state прослеживается слабая зависимость для данных. Для размера тестовых данных видна зависимость на относительной ошибке: чем меньше тестовых данных (то есть чем больше обучающих, test_size меньше), тем ниже эта ошибка. Например для state = 1 при значениях test_size 0.2, 0.3, 0.4 имеем ошибки 32.494, 33.971, 36.591, откуда видна эта зависимость